pkg/lib/_internal/packagemanager-abstract.ts

```diff
@@ -0,0 +1,99 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2025 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <https://www.gnu.org/licenses/>.
+ */
+
+export interface ProgressData {
+    // PackageManager waits on a lock or another operation to finish
+    waiting: boolean
+    // If not null, the operation can be cancelled
+    cancel?: (() => void) | null
+    // FIXME, mistake? should be percentage?
+    absolute_percentage: number
+}
+
+export type ProgressCB = (data: ProgressData) => void;
+export interface InstallProgressData extends ProgressData {
+    // One of the Enum states
+    info: number
+    // Package being downloaded or installed
+    package: string
+}
+
+export type InstallProgressCB = (data: InstallProgressData) => void;
+
+export interface MissingPackages {
+    // Packages that were requested, are currently not installed, and can be installed
+    missing_names: string[]
+    // The full package IDs corresponding to missing_names (the ID different per backend)
+    missing_ids: number[]
+    // Packages that were requested, are currently not installed, but can't be found in any repository
+    unavailable_names: string[]
+    // The to be installed package IDs, temporary kept (the ID different per backend)
+    install_ids: number[]
+    // The full package IDs corresponding to remove_names (the ID different per backend)
+    remove_ids: number[]
+
+    // If unavailable_names is empty, a simulated installation of the missing packages
+    // is done and the result also contains these fields:
+
+    // Packages that need to be installed as dependencies of missing_names
+    extra_names: string[]
+    // Packages that need to be removed
+    remove_names: string[]
+    // Bytes that need to be downloaded
+    download_size: number
+}
+
+// Enum containing PackageManager progress callback info states. These are a
+// subset of the PackageKit Enum, other backends should map them to their own
+// logical state.
+export const Enum = {
+    INFO_DOWNLOADING: 10,
```

Comment:
My suggestion: Don't make this lopsided to PackageKit, but define them as strings or independent constants, and translate them into PackageKit enums in the backend. Then you can land this abstraction in main already, and verify that everything works?

But that's not a veto -- everyone else pretending to be PK should work as well.